BB-1580: Add newrelic monitoring documentation (#465)
open-craft/opencraft: documentation/monitoring.md

Monitoring
------------

> New Relic is a SaaS tool for monitoring and alerting. Ocim supports using `New Relic Synthetics` and `New Relic Alerts` for monitoring the instances.

Author: @lgp171188 Should I copy over the private repo documentation here?

> New Relic is a SaaS tool for monitoring and alerting. Ocim supports using `New Relic Synthetics` and `New Relic Alerts` for monitoring the instances.

Reviewer: @giovannicimolin, the idea is to have all the non-private documentation that is not specific to OpenCraft, here. So if the private repo MR contains a description of this, you can copy just that here.

On Alerts, policies can be created to control the alerting behaviour. Alerts has two main components:
* **Alert Policies**: An alert policy is a group of one or more alert conditions. A policy has two settings that apply to all of its conditions: incident preference and notification channels. A policy should be created before adding conditions to it.
* **Alert Conditions**: its a combination of monitored data source, such as a Synthetics Monitor, and thresholds that define the behavior that will be considered a violation. Synthetics only reports if the pings succeeded or not, so our threshold is to check if the ping failed.

> Each **Alert Policy** contains one or more **Alert Conditions** that can be configured to modify the alerting behaviour.

Author: It looks a lot better! Thanks!

For example, on Open edX instances, we create a alert policy with 1 alert per policy, and add an alert condition for each service (LMS, Studio, Preview).
That way, if any service is down, the monitor will trigger an alert. But if there's more than one service failing pings, only a single alert will be fired, to prevent firing a massive amount of alert emails when an entire instance goes down or there are OVH/AWS network issues.

The image below shows the concept behind Alerts.

![New Relic Alerts](./images/newrelic-alerts.png)

A more in deph explanation of how New Relic Alerts work can be found [here](https://docs.newrelic.com/docs/alerts/new-relic-alerts/getting-started/new-relic-alerts-concepts-workflow).

## How it works on Ocim

> Ocim automatically sets up monitoring for each instance, and creates all the necessary NewRelic resources for alerting. It does that by running `.enable_monitoring()` on an instance during provisioning.

Author: :+1:

Additionally, the monitoring set up methods can be called manually to disable monitoring or rebuild NewRelic assets changed due to changes in configuration.

* `.enable_monitoring()`: Set up resources on New Relic to enable monitoring, if the resources already exist, don't change them. The resources here correspond to a URL to be monitored and a list of addresses to be notified. As long as those do not change, the existing resources will be left unmodified. If the URL is changed, the existing monitors for the old URL are not automatically removed - this is important because we don't want to delete those automatically to avoid issues.

> * `.disable_monitoring()`: Deprovision all resources on New Relics related to that instance, including Alert Policies, Alert Conditions, Notification channels (if not shared) and Synthetics monitors. This is used when an instance is archived and might be useful for debugging.

Reviewer: ```suggestion
* `.disable_monitoring()`: Deprovisions all the New Relic resources related to that instance, including Alert Policies, Alert Conditions, Notification channels (if not shared) and Synthetics monitors. This is used when an instance is archived and might be useful for debugging.
```

## Deploying production instances

> For production instances, some manual setup is required to alert to other notification channels (such as PagerDuty or a email list).

Reviewer: ```suggestion
Ocim automatically creates an email notification channel for each email address in `settings.ADMINS` and adds them to the alert policies of all the active instances. If additional email addresses have to be alerted for the incidents related to an instance, the following manual steps will be required.
```

1. Using Django Admin interface, go to the OpenEdXInstance model and find the instance you want to add monitoring to.
2. Add additional monitoring emails to the `additional_monitoring_emails` field and save the model. The field takes in an array of strings with valid email addresses.
3. Apply the New Relic configuration by redeploying the instance (click on `Spawn new appserver` button on the instance page).

## Troubleshooting

> Most issues with Ocim monitored instances can be fixed by running `.enable_monitoring()`, but in case that's failing you can use the following checklist to help you debug the issue:

Reviewer: > Most issues with Ocim monitored instances can be fixed by running `.enable_monitoring()`

@giovannicimolin, is this always true? :stuck_out_tongue:

Author: No. I will rephrase this.

> 1. Check if the monitored endpoints are working as expected and available from the internet. For LMS, the monitoring address is `/heartbeat?extended`. The return of the page will point out if there's any issue with the instance.

Reviewer: > For LMS, the monitoring address is `/heartbeat?extended`

@giovannicimolin, I don't think this is correct. For the LMS, the monitored address is the instance domain. The `/heartbeat?extended` is for the extended heartbeat which includes the health checks for `celery` and `forum` in addition to the health checks for for Mongo and MySQL included in the standard heartbeat endpoint `/heartbeat`.

Author: @lgp171188 I'll drop the mention to the LMS endpoint

Monitoring
------------

New Relic is a SaaS tool for monitoring and alerting. Ocim supports using `New Relic Synthetics` and `New Relic Alerts` for monitoring the instances.

> On Synthetics, monitors can be created to the check liveness of an endpoint/URL by periodically ping it. If the pre-defined, expected response is not received, the monitor can be configured using `New Relic Alerts` to send out alert notifications through various channels including, but not limited to, email.

Reviewer: ```suggestion
On Synthetics, monitors can be created to the check liveness of an endpoint/URL by periodically pinging it. If the pre-defined, expected response is not received, the monitor can be configured using `New Relic Alerts` to send out alert notifications through various channels including, but not limited to, email.
```

> On Alerts, policies can be created to control the alerting behaviour. Alerts has two main components:

Reviewer: ```suggestion
On New Relic Alerts, policies can be created to control the alerting behaviour. An alert has two main components:
```

> * **Alert Policies**: An alert policy is a group of one or more alert conditions. A policy has two settings that apply to all of its conditions: incident preference and notification channels. The incident preference is used to configure how an incident is created - per error, per condition or per policy and the notification channels are the various modes of sending out alert notifications like email, PagerDuty etc. A policy should be created before adding conditions to it.

Reviewer: @giovannicimolin, a newline is required here for the list formatting to work.

> * **Alert Policies**: An alert policy is a group of one or more alert conditions. A policy has two settings that apply to all of its conditions: incident preference and notification channels. The incident preference is used to configure how an incident is created - per error, per condition or per policy and the notification channels are the various modes of sending out alert notifications like email, PagerDuty etc. A policy should be created before adding conditions to it.

Reviewer: ```suggestion
* **Alert Policy**: An alert policy is a group of one or more alert conditions. A policy has two settings that apply to all of its conditions: incident preference and notification channels. The incident preference is used to configure how an incident is created - per error, per condition or per policy and the notification channels are the various modes of sending out alert notifications like email, PagerDuty etc. A policy should be created before adding conditions to it.
```

Each **Alert Policy** contains one or more **Alert Conditions** that can be configured to modify the alerting behaviour.

For example, on Open edX instances, we create a alert policy with 1 alert per policy, and add an alert condition for each service (LMS, Studio, Preview).
That way, if any service is down, the monitor will trigger an alert. But if there's more than one service failing pings, only a single alert will be fired, to prevent firing a massive amount of alert emails when an entire instance goes down or there are OVH/AWS network issues.

The image below shows the concept behind Alerts.

![New Relic Alerts](./images/newrelic-alerts.png)

> A more in deph explanation of how New Relic Alerts work can be found [here](https://docs.newrelic.com/docs/alerts/new-relic-alerts/getting-started/new-relic-alerts-concepts-workflow).

Reviewer: ```suggestion
A more in-depth explanation of how New Relic Alerts work can be found [here](https://docs.newrelic.com/docs/alerts/new-relic-alerts/getting-started/new-relic-alerts-concepts-workflow).
```

> ![New Relic Alerts](./images/newrelic-alerts.png)

Reviewer: @giovannicimolin, you missed adding the image to the PR.

## How it works on Ocim

Ocim automatically sets up monitoring for each instance, and creates all the necessary NewRelic resources for alerting. It does that by running `.enable_monitoring()` on an instance during provisioning.

Additionally, the monitoring set up methods can be called manually to disable monitoring or rebuild NewRelic assets changed due to changes in configuration.

> * `.enable_monitoring()`: Set ups resources on New Relic to enable monitoring, if the resources already exist, overwrite them with new settings.

Reviewer: @giovannicimolin, it won't overwrite any resources. It uses the database to check if resources exist and does nothing if they do. Otherwise, it creates a fresh set of resources and doesn't check the New Relic API for any existing resources for the instance.

Author: @lgp171188 So it won't update existing resources if settings are changed?

Reviewer: @giovannicimolin, the resources here correspond to a URL to be monitored and a list of addresses to be notified. As long as those do not change, the existing resources will be left unmodified. If the URL is changed, the existing monitors for the old URL are not automatically removed - this is important because we don't want to delete those automatically to avoid issues (cf. [this comment](https://github.com/open-craft/opencraft/commit/7fb071b4cf0e06d275094839f7c60a0d0f4f8d7c#diff-c6922995a704923a2f6da744ec24a91aR94))

Author: Ok, I added this to the documentation

* `.disable_monitoring()`: Deprovision all resources on New Relics related to that instance, including Alert Policies, Alert Conditions, Notification channels (if not shared) and Synthetics monitors. This is used when an instance is archived and might be useful for debugging.

## Deploying production instances

For production instances, some manual setup is required to alert to other notification channels (such as PagerDuty or a email list).

> 1. Using Django Shell (`make shell`), find the instance you want to add monitoring to.

Reviewer: @giovannicimolin, this can be done using the Django admin interface.

Reviewer: @giovannicimolin, it needs a redeployment. We always do a redeployment to apply any changes made to an instance's configuration. This might be a lot of work than running `enable_monitoring()`, but it is much safer because there always logs that capture what was done.

Author: :+1:

> * `.enable_monitoring()`: Sets up resources on New Relic to enable monitoring, if the resources already exist, don't change them. The resources here correspond to a URL to be monitored and a list of addresses to be notified. As long as those do not change, the existing resources will be left unmodified. If the URL is changed, the existing monitors for the old URL are not automatically removed - this is important because we don't want to delete those automatically, to avoid issues.

Reviewer: ```suggestion
* `.enable_monitoring()`: Sets up resources on New Relic to enable monitoring, if the resources already exist, it doesn't change them. The resources here correspond to the URLs to be monitored and a list of email addresses to be notified. As long as those do not change, the existing resources will be left unmodified. If the URL is changed, the existing monitors for the old URL are not automatically removed - this is important because we don't want to delete those automatically, to avoid issues.
```

> 1. Using Django Admin interface, go to the OpenEdXInstance model and find the instance you want to add monitoring to.

Reviewer: ```suggestion
1. Using the Django Admin interface, go to the `OpenEdXInstance` model and find the instance you want to update the monitoring addresses of.
```

> 2. Add additional monitoring emails to the `additional_monitoring_emails` field and save the model. The field takes in an array of strings with valid email addresses.

Reviewer: ```suggestion
2. Add the email addresses to the `additional_monitoring_emails` field and save the model. The field takes in an array of strings with valid email addresses.
```

So if any service is down, the monitor will create an incident and trigger an alert. But if more than one service is failing on an instance, only a single incident will be created and a single alert notification will be sent due to the per-policy incident preference. This prevents the creation a large number of incidents and the resulting alert notifications when an instance goes down or there are network issues.

The image below shows the concept behind Alerts.

![New Relic Alerts](./images/newrelic-alerts.png)

A more in-depth explanation of how New Relic Alerts work can be found [here](https://docs.newrelic.com/docs/alerts/new-relic-alerts/getting-started/new-relic-alerts-concepts-workflow).

## How it works on Ocim

Ocim automatically sets up monitoring for each instance, and creates all the necessary NewRelic resources for alerting. It does that by executing the `.enable_monitoring()` method on an instance after it is activated.

Additionally, the `.enable_monitoring()` and `.disable_monitoring()` methods of an instance can be called manually from the Django shell to update the New Relic resources on configuration changes or to disable monitoring.

> * `.enable_monitoring()`: Sets up resources on New Relic to enable monitoring, if the resources already exist, don't change them. The resources here correspond to a URL to be monitored and a list of addresses to be notified. As long as those do not change, the existing resources will be left unmodified. If the URL is changed, the existing monitors for the old URL are not automatically removed - this is important because we don't want to delete those automatically, to avoid issues.

Reviewer: ```suggestion
* `.enable_monitoring()`: Sets up resources on New Relic to enable monitoring, if the resources already exist, it doesn't change them. The resources here correspond to a URL to be monitored and a list of addresses to be notified. As long as those do not change, the existing resources will be left unmodified. If the URL is changed, the existing monitors for the old URL are not automatically removed - this is important because we don't want to delete those automatically, to avoid issues.
```
